tools/codegen/api/types.py

```diff
@@ -157,16 +157,23 @@ def name(self) -> str:
         return n
 
     # Render the C++ declaration for this signature
-    def decl(self) -> str:
+    def decl(self, *, prefix: str = "", is_redispatching_fn: bool = False) -> str:
         returns_type = cpp.returns_type(self.func.returns)
-        cpp_args_str = ', '.join(a.decl() for a in self.arguments())
-        return f"{returns_type} {self.name()}({cpp_args_str})"
+        cpp_args = [a.decl() for a in self.arguments()]
+        if is_redispatching_fn:
+            cpp_args = ['c10::DispatchKeySet dispatchKeySet'] + cpp_args
+        cpp_args_str = ', '.join(cpp_args)
+        name = prefix + self.name()
+        return f"{returns_type} {name}({cpp_args_str})"
```

Comment:
The change here looks a bit questionable, because it is not clear to me that `is_redispatching_fn` shouldn't be an intrinsic property of the entire `CppSignature` itself, similar to how `faithful` is a property. I suppose one argument for not making it intrinsic is that we don't currently have a concept of DispatchKeySet in Argument schema (and you'd really want one, because `arguments` had better report this.)

Reply:
Making it a property of the `CppSignature` makes sense to me, but I don't think it plays out as cleanly if we want to keep the C++ API and redispatch API in two separate .cpp files- since right now, calls to `ComputeFunction` are tied to a single file. E.g. I could update `ComputeFunction` to write out the faithful, non-faithful, and redispatching definition of every op that you pass into it, but right now they'll all get written to the same file.

Is it worth updating `gen.py` to do sharding, similar to what tracing and autograd do? Then I could write each (faithful, non-faithful, redispatch) trio together, but split up the ops among different files to keep the max file size down.

Reply:
So, if I understand you correctly, the fear is related to the combinatorial explosion where you need each faithful variant and each redispatch variant, if you wrote these all inline. But I don't think you have to do that; you can still keep the top level structure you have (which is pass in a property that says dispatch or not), and then just construct the full CppSignatureGroup as redispatch or not based on this property.

Reply:
No, I was more thinking that "I'll need this top-level `is_redispatching` flag either way in order to redirect normal functions to one file and redispatching variants to another, so I may as well keep all of the logic in the same place", although some of the logic leaked into `types.py` here. Another option though that would let me kill the `is_redispatching_fn` flag in ComputeFunction would be to have it always generate both the normal and redispatching variants of each function (like what we do with the faithful/nonfaithful API right now), and just shard the output across multiple files to avoid clogging up `Functions.cpp`.

I'm going to try what you're saying for now though and see how it turns out. One downside worth calling out (that I also want to confirm) - I'd need to add DispatchKeySet as a jit type that can be translated to cpp. That's a little weird, because we also have explicit logic around the Dispatcher now to say "if we see a kernel with a first argument of type DispatchKeySet, let's hide it from jit" (to avoid having to make it convertible to/from an ivalue, etc)

Reply:
> I may as well keep all of the logic in the same place

I generally see this positively. For example, the reason why it's the same function that does declarations and definitions (rather than two separate functions) is to encourage code sharing between the two, because the underlying structure is the same.

> One downside worth calling out (that I also want to confirm) - I'd need to add DispatchKeySet as a jit type that can be translated to cpp. That's a little weird, because we also have explicit logic around the Dispatcher now to say "if we see a kernel with a first argument of type DispatchKeySet, let's hide it from jit" (to avoid having to make it convertible to/from an ivalue, etc)

Yes, I agree this is the weird thing. One possibility is that `Argument` field on `Binding` should be optional; if that works out that's a clue that this model does make sense given the code.

Reply:
(To be explicit, I don't think faking up an Argument JIT type for dispatch key set; it doesn't correspond to JIT schema!)

Reply:
Yeah, I'm right now I'm more inclined towards the current state of the PR (patching `DispatchKeySet` in as an extra argument in the two places I need it, and explicitly not putting in the model). I don't think that making `Argument` an optional field on `Binding` helps much, since you still need to create an `Argument` representing the DispatchKeySet at some point. Just to lay it out, here I what I see as the options:

1) The current state: an `is_redispatching` flag tacked onto `ComputeFunction` that basically tells it whether or not to patch in a `DispatchKeySet` argument on top of whatever model produced. Totally keep `DispatchKeySet` out of the model.
2) Fully add DispatchKeySet to the model: make it a part `Arguments`, tying it directly to the jit `FunctionSchema` type. Give `CppSignature` an `is_redispatching` flag, similar to `faithful`, that basically tells it whether or not to include the `DispatchKeySet` argument in its signature. This would also require hardcoding the extra DispatchKeySet argument into `FunctionSchema.parse`, since it doesn't actually live in native_functions.yaml. The benefit is that the top-level calls from `ComputeFunction` can just pass in an `is_redispatching` flag and everything works.
3) We could just do everything in `CppSignature`- i.e. still have an `is_redispatch` flag on `CppSignature` to decide whether or not to include a DispatchKeySet arg, but just tack it on there instead of getting it from the model.
4) Do the same thing as (2), but avoid giving `ComputeFunction` a new `is_redispatching` flag. We can avoid it by having ComputeFunction always generate both normal/redispatching variants, and just shard the output across files to avoid `Functions.cpp` from getting too big.


To me all of these variants end up conjuring up a new DispatchKeySet arg somewhere, and the question is just whether or not to do it at the top-level, or plumb a flag through the system that eventually creates it deeper in the model. The fact that `DispatchKeySet` should be hidden from jit makes me lean more towards doing everything at the top level (the current state of the PR), with a better comment explaining why.

Reply:
Agreed! And it will not be hard to fix this later if we do change our minds.